Create TextureStream.md (#2743)
MicrosoftEdge/WebView2Feedback: specs/APIReview_TextureStream.md

prevents apps from rendering real-time video inside the web contents. This forces apps to
render the real-time video on top of the web contents, which is limiting. Rendering video on 
top constrains the user experience and it may also cause performance problems.  
We can ask the native apps to use web renderer for video handling because web standard already 
provides these features through WebRTC APIs. The end developers, however, prefer to use 
their existing engine such as capturing and composition, meanwhile using WebRTC API for rendering.

# Description
The proposed APIs will allow the end developers to stream the captured or composed video frame to 
the WebView renderer where Javascript is able to insert the frame to the page through W3C standard 
API of Video, MediaStream element for displaying it.
The API will use the shared GPU texture buffer so that it can minimize the overall cost with 
regards to frame copy.

# Examples
Javascript
// User click the video capture button.
document.querySelector('#showVideo').addEventListener('click',
  e => getStreamFromTheHost(e));
async function getStreamFromTheHost(e) {
  try {
    // Request stream to the host with unique stream id.

>     const stream = await window.chrome.webview.getTextureStream('webview2-abcd1234'); 

Reviewer: This sample doesn't demonstrate JS handling stop and restarting (is it via active/inactive events?). Is that expected or does it not explicitly handle that?

>     const stream = await window.chrome.webview.getTextureStream('webview2-abcd1234'); 

Reviewer: Is the only scenario for this API to have native sending video to JS via MediaStream to render in web content? Maybe I'm thinking of a different scenario where the JS side could edit or create video to send back to native.

Author: getTextureStream returns W3C standard of MediaStream/MediaStreamTtrack objects and developer can call commands/events on them as documented on W3C spec. 

Yes, one of things they can do with these objects is to edit (insertable) the video frame. However, it is also W3C standard, no special things for this API.

Do you think we need to add it even though there is no difference (https://web.dev/mediastreamtrack-insertable-media-processing/)?

Author: Insertable stream API (W3C standard API) is the one that allow editing texture stream, but it is used for another API of 'registerTextureStream', which will come on next round. (It is under design review, and shortly on PR in this week).

>     const stream = await window.chrome.webview.getTextureStream('webview2-abcd1234'); 

Reviewer: A bunch of my comments are applicable to the API surface not the sample code and are left unresolved. Can you look through them and reply to them please?

  [hWnd](ICoreWebView2StagingTextureStream* sender, ICoreWebView2StagingTextureStreamTextureErrorEventArgs* args) {
    COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND kind;
    HRESULT hr = args->get_Kind(&kind);
    assert(SUCCEEDED(hr));
    switch (kind)
    {
    case COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED:
    case COREWEBVIEW2_TEXTURE_STREAM_ERROR_BUFFER_NOT_FOUND:
    case COREWEBVIEW2_TEXTURE_STREAM_ERROR_BUFFER_IN_USE:
      // assert(false);
      break;
    default:
      break;
    }
    return S_OK;
  }).Get(), &texture_token);
    
// TextureStream APIs are called in the UI thread on the WebView2 process meanwhile Video capture
// and composition could happen in worker thread or out of process.
LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
{
  static ICoreWebView2Staging3* webview2_17 = nullptr;
  TCHAR greeting[] = _T("Hello, Windows desktop!");
  ComPtr<ID3D11Device> d3d_device;
  HANDLE slimCoreHandle;
  HRESULT hr;
  ComPtr<ICoreWebView2StagingTexture> texture_buffer;
  int64_t bufferId = -1;
  switch (message)
  {

>   case IDC_TEST_SEND_TEXTURE:

Reviewer: Where do these window messages come from? If we don't perform this part of the sample in a different process, would it make the sample smaller and easier to understand?

Author: I've updated it to be simpler, but let me know if you think it needs more changes.

[uuid(96c27a45-f142-4873-80ad-9d0cd899b2b9), object, pointer_default(unique)]
interface ICoreWebView2Staging3 : IUnknown {
  /// Registers the stream id that the host can handle, providing a
  /// texture stream when requested from the WebView2's JavaScript code.
  /// The host can register multiple unique stream instances, each with
  /// a unique stream ID, enabling the host to stream from different sources
  /// concurrently.
  /// The host should call this only once for unique streamId. The second
  /// call of already created streamId without destroying
  /// ICoreWebView2StagingTextureStream object will return an error.
  /// 'd3dDevice' is used for creating shared IDXGI resource and NT shared
  /// of it. The host should use Adapter of the LUID from the GetRenderAdapterLUID
  /// for creating the D3D Device.
  HRESULT CreateTextureStream(
      [in] LPCWSTR streamId,
      [in] IUnknown* d3dDevice,
      [out, retval ] ICoreWebView2StagingTextureStream** value);
  /// Get the graphics adapter LUID of the renderer. The host should use this
  /// LUID adapter when creating D3D device to use with CreateTextureStream().

> HRESULT GetRenderAdapterLUID([out, retval] LUID* luid);

Reviewer: This can be a property. 
```suggestion
[propget] HRESULT RenderAdapterLUID([out, retval] LUID* value);
```

  /// Listens for change of graphics adapter LUID of the browser.
  /// The host can get the updated LUID by GetRenderAdapterLUID. It is expected
  /// that the host updates texture's d3d Device with UpdateD3DDevice,
  /// removes existing buffers and creates new buffer.

>   HRESULT add_RenderAdapterLUIDUpdated(

Reviewer: Events for property changes are named [PropertyName]Changed.

```suggestion
  HRESULT add_RenderAdapterLUIDChanged(
```

  /// must be registered first in order for the request to succeed.
  /// The added filter will be persistent until
  /// ICoreWebView2StagingTextureStream is destroyed or
  /// RemoveRequestedFilter is called.
  /// The renderer does not support wildcard so it will compare
  /// literal string input to the requesting frame origin. So, the input string
  /// should have a scheme like https://.
  /// For example, https://www.valid-host.com, http://www.valid-host.com are
  /// valid origins but  www.valid-host.com, or *.valid-host.com. are not
  /// valid origins.
  /// getTextureStream() will fail unless the requesting frame's origin URL is
  /// added to the request filter.
  HRESULT AddRequestedFilter([in] LPCWSTR origin);
  /// Remove added origin, which was added by AddRequestedFilter.
  HRESULT RemoveRequestedFilter([in] LPCWSTR origin);
  /// Listens for stream requests from the Javascript's getTextureStream call
  /// for the given stream id. It is called for the first request only, the
  /// subsequent requests of same stream id will not be called.
  /// It is expected that the host provides the stream within 10s after

>   /// being requested. The first call to Present() fulfills the stream request.

Reviewer: What does it mean to fulfill the stream request? Is the JS getTextureStream call waiting until the first Present call before it returns the MediaStream?

      [in] ICoreWebView2StagingTextureStreamStopRequestedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for stop stream request.
  HRESULT remove_StopRequested(
      [in] EventRegistrationToken token);
  /// Creates shared buffer that will be referenced by the host and the browser.
  /// By using the shared buffer mechanism, the host does not have to
  /// send respective texture to the renderer, instead it notifies it
  /// with internal buffer id, which is the identity of the shared buffer.
  /// The shared buffer is 2D texture, IDXGIResource, format and will be
  /// exposed through shared HANDLE or IUnknown type through ICoreWebView2StagingTexture.
  /// Whenever the host has new texture to write, it should ask
  /// reusable ICoreWebView2StagingTexture from the GetAvailableBuffer,
  /// which returns ICoreWebView2StagingTexture.
  /// If the GetAvailableBuffer returns an error, then the host calls the
  /// CreateBuffer to allocate new shared buffer.
  /// The API also registers created shared handle to the browser once it
  /// created the resource.
  HRESULT CreateBuffer(
    [in] UINT32 width,
    [in] UINT32 height,

>     [out, retval] ICoreWebView2StagingTexture** buffer);

Reviewer: You call it a buffer in the parameter name and in the Create method name, but the type name is Texture. How about making the type name and Create name match
 * TextureStreamBuffer?
 * TextureStreamTexture?
 * TextureStreamFrame?

>   /// GetAvailableBuffer can be called on any thread like SetBuffer.

Reviewer: Great note that this can be called on any thread. Please make sure all methods and properties that work on other threads are noted as working on other threads. Its stated generally that WebView2 methods only work on the WebView2 UI thread so we need to explicitly call out anything that doesn't work like that.

>   HRESULT GetAvailableBuffer([out, retval] ICoreWebView2StagingTexture** buffer);

Reviewer: What does GetAvailableBuffer do? Its documentation above needs to explain.

  /// Remove texture buffer when the host removes the backed 2D texture.
  /// The host can save the existing resources by deleting 2D textures
  /// when it changes the frame sizes.
  HRESULT RemoveBuffer([in] ICoreWebView2StagingTexture* buffer);
  /// Indicates that the buffer is ready to present.
  /// The buffer must be retrieved from the GetAvailableBuffer.
  /// The host writes new texture to the local shared 2D texture of
  /// the buffer id, which is created via CreateBuffer.
  /// SetBuffer API can be called in any thread.
  HRESULT SetBuffer([in] ICoreWebView2StagingTexture* buffer,
    [in] ULONGLONG timestamp);
  /// Render texture that is current set ICoreWebView2StagingTexture.

>   HRESULT Present();

Reviewer: Why is SetBuffer and Present different methods? Why doesn't Present take the parameters of SetBuffer and present them?

  /// Stop streaming of the current stream id.
  /// API calls of Present, CreateBuffer will fail after this
  /// with an error of COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED.
  /// The Javascript can restart the stream with getTextureStream.
  HRESULT Stop();
  /// Event handler for those that occur at the Renderer side, the example

>   /// are CreateBuffer, Present, or Stop.

Reviewer: This comments doesn't seem to apply to add_TextureError does it?

>   HRESULT add_TextureError(

Reviewer: Event names should be verb phrases like ErrorOccurred or ErrorDetected

>   HRESULT add_TextureError(

Reviewer: If these are general TextureStream errors then we don't need the 'Texture' name prefix.

      [in] ICoreWebView2StagingTextureStreamTextureErrorEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for texture error event.
  HRESULT remove_TextureError([in] EventRegistrationToken token);
  /// Updates d3d Device when it is updated by RenderAdapterLUIDUpdated
  /// event.

>   HRESULT UpdateD3DDevice([in] IUnknown* d3dDevice);

Reviewer: What does update mean in this context? Is it obvious to someone familiar with D3D what Update would mean here? Can you be more specific in this comment about what Update means?

  /// Called to provide the implementer with the event args for the
  /// corresponding event. There are no event args and the args
  /// parameter will be null.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);
}
/// This is the callback for texture stream rendering error.
[uuid(52cb8898-c711-401a-8f97-3646831ba72d), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamTextureErrorEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] ICoreWebView2StagingTextureStreamTextureErrorEventArgs* args);
}
/// This is the event args interface for texture stream error callback.
[uuid(0e1730c1-03df-4ad2-b847-be4d63adf700), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamTextureErrorEventArgs : IUnknown {
  /// Error kind.
  [propget] HRESULT Kind([out, retval]
      COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND* value);
  // Texture buffer that the error is associated with.

>   HRESULT GetBuffer([out, retval] ICoreWebView2StagingTexture** buffer);

Reviewer: This can be a property instead right? `[propget] HRESULT Buffer([out, retval] ICoreWebView2StagingTexture** value);`

The API will use the shared GPU texture buffer so that it can minimize the overall cost with
regards to frame copy.

# Examples

## Javascript

This is Javascript code common to both of the following samples:

```js
// getTextureStream sample.

// Developer can stream from the host and its returned object is MediaStream
// that is same what getUserMedia returns. Technically, developer can do all
// what MediaStream and its video MediaStreamTrack provide.

// Scenario: User clicks the button and show stream from the host via video element.
document.querySelector('#showVideo').addEventListener('click',
  e => getStreamFromTheHost(e));
async function getStreamFromTheHost(e) {
  try {
    // Request stream to the host with unique stream id.
    // getTextureStream W3C standard MediaStream object that has only video
    // MediaStreamTrack.
    const stream = await window.chrome.webview.getTextureStream('webview2-abcd1234');
    // The MediaStream object is returned and it gets video MediaStreamTrack element from it.
    const video_tracks = stream.getVideoTracks();
    const videoTrack = video_tracks[0];
    window.videoTrack = videoTrack;
    // Show the video via Video Element.

>     document.getElementById(video_id).srcObject = stream;

Author: I will create video element instead of adding HTML that has 'video_id'.

  /// But once they call Present, the buffer becomes "in use" until
  /// they call SetBuffer and Present on a different buffer and wait a bit
  /// for the original buffer to stop being used.
  COREWEBVIEW2_TEXTURE_STREAM_ERROR_BUFFER_IN_USE,
} COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND;
/// This is ICoreWebView2Staging3 that returns the texture stream interface.
[uuid(96c27a45-f142-4873-80ad-9d0cd899b2b9), object, pointer_default(unique)]
interface ICoreWebView2Staging3 : IUnknown {
  /// Registers the stream id that the host can handle, providing a
  /// texture stream when requested from the WebView2's JavaScript code.
  /// The host can register multiple unique stream instances, each with
  /// a unique stream ID, enabling the host to stream from different sources
  /// concurrently.
  /// The host should call this only once for unique streamId. The second
  /// call of already created streamId without destroying
  /// ICoreWebView2StagingTextureStream object will return an error.
  /// 'd3dDevice' is used for creating shared IDXGI resource and NT shared
  /// of it. The host should use Adapter of the LUID from the GetRenderAdapterLUID
  /// for creating the D3D Device.

>   HRESULT CreateTextureStream(

Author: yes, your understanding is correct.

>   HRESULT CreateTextureStream(

Author: - TextureStream object should be video source, which is like single camera in the Chromium engine perspective. So, it isn't doable for single TextureStream object to have multiple names; The name is like device name. When the host supports multiple names, then it has to create multiple TextureStream that will be clear even with the event, otherwise every events 

- CreateTextureStream needs D3DDevice object, which can be created from LUID that the host can get from GetRenderAdapterLUID.   GetRenderAdapterLUID API can successfully return a value only when GPU process is created that usually happens in navigation.  

- TextureStream is associated with its counterpart in the Javascript, getTextureStream and registerTextureStream. These Javascript APIs can be called only in the page (window) so it looks more natural for it to be in CoreWindow2.
-

>   HRESULT CreateTextureStream(

Reviewer: From discussion, by the scenario we need one texturestream to be able to work with multiple CoreWebView2s. Since the texturestream isn't tied to one CoreWebView2 we should move these methods to the CoreWebView2Environment. On the environment you should be able to CreateTextureStream. But as you pointed out there is no connection required to particular CoreWebView2 objects, so there's no Add or Register method necessary. Just create the TextureStream with a particular name and then its available by that name in all the CoreWebView2s for that environment.

We didn't talk as much about the RenderAdapterLUID property or the RenderAdapterLUIDChanged event, but it sounds like maybe those also should move to the CoreWebView2Environment.

      [in] LPCWSTR streamId,
      [in] IUnknown* d3dDevice,
      [out, retval ] ICoreWebView2StagingTextureStream** value);
  /// Get the graphics adapter LUID of the renderer. The host should use this
  /// LUID adapter when creating D3D device to use with CreateTextureStream().
  [propget] HRESULT RenderAdapterLUID([out, retval] LUID* luid);
  /// Listens for change of graphics adapter LUID of the browser.
  /// The host can get the updated LUID by RenderAdapterLUID. It is expected
  /// that the host updates texture's d3d Device with UpdateD3DDevice,
  /// removes existing buffers and creates new buffer.
  HRESULT add_RenderAdapterLUIDChanged(
      [in] ICoreWebView2StagingRenderAdapterLUIDChangedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for start stream request.
  HRESULT remove_RenderAdapterLUIDChanged(
      [in] EventRegistrationToken token);
}
/// This is the interface that handles texture streaming.

> /// The most of APIs have to be called on UI thread.

Author: sounds good.

  /// The event is triggered when all requests for given stream id closed
  /// by the Javascript, or the host's Stop API call.
  /// TextureBuffer related API calls after this event will return an error
  /// of HRESULT_FROM_WIN32(ERROR_INVALID_STATE).
  HRESULT add_StopRequested(
      [in] ICoreWebView2StagingTextureStreamStopRequestedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for stop stream request.
  HRESULT remove_StopRequested(
      [in] EventRegistrationToken token);
  /// Creates TextureBuffer that will be referenced by the host and the browser.
  /// By using the TextureBuffer mechanism, the host does not have to
  /// send respective texture to the renderer, instead it notifies it
  /// with internal TextureBuffer id, which is the identity of the TextureBuffer.
  /// The TextureBuffer is 2D texture, IDXGIResource, format and will be
  /// exposed through shared HANDLE or IUnknown type through ICoreWebView2StagingTexture.
  /// Whenever the host has new texture to write, it should ask
  /// reusable ICoreWebView2StagingTexture from the GetAvailableBuffer,
  /// which returns ICoreWebView2StagingTexture.

>   /// If the GetAvailableBuffer returns an error, then the host calls the

Author: No, the host doesn't have to call GetAvailableBuffer if the host sure that the machine has enough resources. (They can create 10 buffers (CreateBuffers) and keep cycling).

  /// CreateBuffer to allocate new TextureBuffer.
  /// The API also registers created shared handle to the browser once it
  /// created the resource.

  /// Unit for width and height is texture unit (in texels).
  /// 'https://learn.microsoft.com/en-us/windows/win32/api/d3d11ns-d3d11-d3d11_texture2d_desc'
  HRESULT CreateBuffer(
    [in] UINT32 width,
    [in] UINT32 height,
    [out, retval] ICoreWebView2StagingTexture** buffer);
  /// Returns reuseable TextureBuffer for video frame rendering.
  /// Once the renderer finishes rendering of TextureBuffer's video frame, which
  /// was requested by Present, the renderer informs the host so that it can
  /// be reused. The host has to create new TextureBuffer with CreateBuffer

>   /// if the API return an error HRESULT_FROM_WIN32(ERROR_NO_MORE_ITEMS).

Author: It is up to the system, usually from 3 to 5 buffers. If the host sends image (call Present) with 120 fps, but the renderer is slow, then it could go more than 10 buffers.   If the host can't create more buffers, then it can remove buffer with RemoveBuffer of old frame even though it might not fully be rendered yet in the browser. This is a mechanism already used in the Chromium engine for camera capture.

  HRESULT GetAvailableBuffer([out, retval] ICoreWebView2StagingTexture** buffer);
  /// Remove TextureBuffer when the host removes the backed 2D texture.
  /// The host can save the existing resources by deleting 2D textures
  /// when it changes the frame sizes.
  HRESULT RemoveBuffer([in] ICoreWebView2StagingTexture* buffer);

>   /// Indicates that the TextureBuffer is ready to present.

Reviewer: Shouldn't Present indicate that the buffer is ready to present?

Author: change to 
/// Sets rendering image/resource through ICoreWebView2StagingTexture.

>   /// The TextureBuffer must be retrieved from the GetAvailableBuffer.

Reviewer: Or created via CreateBuffer right?

Author: yes

  /// The host writes new texture to the local shared 2D texture of
  /// the TextureBuffer id, which is created via CreateBuffer.
  /// SetBuffer API can be called in any thread.

  /// `timestamp` is video capture time with unit of microseconds units.
  /// The value does not have to be exact captured time, but it should be
  /// increasing order, the next Present's TextureBuffer should have later

>   /// time.

Author: renderer (composition) ignores incoming video frame (texture) if its timestamp is equal or prior to the current compositing video frame.

>   HRESULT SetBuffer([in] ICoreWebView2StagingTexture* buffer,

Reviewer: SetBuffer sounds like it is setting a property. Like there is one buffer member inside the texture stream and when you call SetBuffer its changing that one member to point to the new provided parameter. Is that what is happening? If SetBuffer is adding to a list of buffers then it should be called AddBuffer (and the RemoveBuffer method is well named Add/Remove go together generally). If this is SetBuffer, then RemoveBuffer should probably be called ClearBuffer, and it wouldn't need to take a parameter.

Author: SetBuffer is single item only so the latest SetBuffer will override the previous ones.

RemoeBuffer is actually removing a TextureBuffer. I will change a name to RemoveTextureBuffer that will remove a buffer from the browser side (The browser received and kept it when the host call Present).

    [in] UINT64 timestamp);

>   /// Render texture that is current set ICoreWebView2StagingTexture.

Author: the most recent call. will change comment to 
"
  /// Render ICoreWebView2StagingTexture resource, which is the most recent
  /// call to SetBuffer.
"

      return S_OK;
    }).Get(), &stopped_token));
}  // CreateTextureStream

HRESULT CallWebView2API(bool createBuffer,
                        UINT32 width,
                        UINT32 height,
                        bool requestAvailableBufer,
                        bool sendTexture,
                        ICoreWebView2StagingTexture* send_texture,
                        UINT64 timestamp,
                        ICoreWebView2StagingTexture** texture) {
  if (createBuffer) {
    // Create TextureBuffer.
    return webviewTextureStream->CreateSharedBuffer(width, height, &texture);
  }

  if (requestAvailableBufer) {
    return webviewTextureStream->GetAvailableBuffer(&texture);
  }

  if (sendTexture) {
    // Notify the renderer for updated texture on the TextureBuffer.
    webviewTextureStream->SetBuffer(send_texture, timestamp);
    webviewTextureStream->Present();
  }
}
```

# API Details
```
[v1_enum]
typedef enum COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND {
  /// The host can't create a TextureStream instance more than once
  /// for a specific stream id.
  COREWEBVIEW2_TEXTURE_STREAM_ERROR_STREAM_ID_ALREADY_REGISTERED,

>   /// Occurs when the host calls CreateBuffer or Present

Author: based on conversion, remove COREWEBVIEW2_TEXTURE_STREAM_ERROR_STREAM_ID_ALREADY_REGISTERED and update commented on with COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED.

  /// Sets rendering image/resource through ICoreWebView2StagingTexture.
  /// The TextureBuffer must be retrieved from the GetAvailableBuffer or
  /// created via CreateBuffer.
  /// It is expected that hhe host writes new image/resource to the local
  /// shared 2D texture of the TextureBuffer (handle/resource).
  /// SetBuffer API can be called in any thread.

  /// `timestampInMs` is video capture time with microseconds units.
  /// The value does not have to be exact captured time, but it should be
  /// increasing order because renderer (composition) ignores incoming
  /// video frame (texture) if its timestampInMs is equal or prior to
  /// the current compositing video frame.
  HRESULT SetBuffer([in] ICoreWebView2StagingTexture* buffer,
    [in] UINT64 timestampInMs);
  /// Render ICoreWebView2StagingTexture resource, which is the most recent
  /// call to SetBuffer.
  HRESULT Present();
  /// Stop streaming of the current stream id.
  /// API calls of Present, CreateBuffer will fail after this
  /// with an error of COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED.
  /// The Javascript can restart the stream with getTextureStream.

>   HRESULT Stop();

Author: updated

  HRESULT remove_WebTextureReceived([in] EventRegistrationToken token);
  /// Event handler for stopping of the receiving texture stream.
  /// It is expected that the host releases any holding handle/resource from
  /// the WebTexture before an event handler returns.
  HRESULT add_WebTextureStreamStopped(
      [in] ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for WebTextureStreamStopped event.
  HRESULT remove_WebTextureStreamStopped([in] EventRegistrationToken token);
}
/// TextureBuffer that the host writes to so that the Renderer
/// will render on it.
[uuid(0836f09c-34bd-47bf-914a-99fb56ae2d07), object, pointer_default(unique)]
interface ICoreWebView2StagingTexture : IUnknown {
    /// Returns shared Windows NT handle. The caller expected to open it with
    /// ID3D11Device1::OpenSharedResource1 and writes the incoming texture to it.
    [propget] HRESULT Handle([out, retval] HANDLE* handle);
    /// Returns IUnknown type that could be query interface to IDXGIResource.
    /// The caller can write incoming texture to it.

>     [propget] HRESULT Resource([out, retval] IUnknown** resource);

Author: add [propput] Timestamp here, and remove it from the PresentBuffer.

  /// handle value as a unique TextureBuffer key.
  /// If the host opens its own resources by handle, then it is suggested
  /// that the host removes those resources when the handle's texture size
  /// is changed because the browser also removed previously allocated different
  /// sized buffers when image size is changed.
  [propget] HRESULT Handle([out, retval] HANDLE* handle);

  /// TextureBuffer resource.
  /// The same resource value will be used for same TextureBuffer so the host can use
  /// resource value as a unique TextureBuffer key.
  /// ICoreWebView2StagingTextureStream object has a reference of the resource
  /// so ICoreWebView2StagingWebTexture holds same resource object for
  /// the same TextureBuffer.
  [propget] HRESULT Resource([out, retval] IUnknown** resource);

  /// It is timestamp of the web texture. Javascript can set this value
  /// with any value, but it is suggested to use same value of its original
  /// video frame that is a value of SetBuffer so that the host is able to
  /// tell the receiving texture delta.
  [propget] HRESULT Timestamp([out, retval] UINT64* value);
}

> ```

Reviewer: We've updated the API spec template to have a section for TypeScript that defines the JavaScript APIs you're adding. It would be good to add that.

Reviewer: Try the following out in https://www.typescriptlang.org/play to make sure it looks like valid TS and looks like you expect
```

---ts
interface WebView extends EventTarget {
    // ... leaving out existing methods

    /// documentation...
    getTextureStream(textureStreamId: string): MediaStream;

    /// documentation...
    registerTextureStream(textureStreamId: string, textureStream VideoStreamTrack): void;

    /// documentation...
    unregisterTextureStream(textureStreamId: string): void;
}

---

```

# Description
The proposed APIs will allow the end developers to stream the captured or composed video frame to the WebView renderer where Javascript is able to insert the frame to the page through W3C standard API of Video, MediaStream element for displaying it.
The API will use the shared GPU texture buffer so that it can minimize the overall cost with regards to frame copy.

The proposed APIs have dependency on the DirectX and its internal attributes such as
adapter LUID so it supports only Win32/C++ APIs at this time.

# Examples

## Javascript

> This is Javascript code common to both of the following samples:

Reviewer: Please add two subheadings and make full copies of the code for each

### Render video stream in video HTML element

In this sample, the native code generates video in a texture stream, it is sent to JavaScript, which renders it into a video HTML element.

... js
... c++

### Edit video in JavaScript and send back to native

In this sample, the native code generates video in a texture stream, it is sent to JavaScript, JavaScript edits the video, and sends it back to native code.

... js
... c++

    if (FAILED(hr))
      return hr;

    hr = webviewTextureStream->GetAvailableBuffer(&textureBuffer);
    assert(SUCCEEDED(hr));
  }

  ComPtr<IUnknown> dxgiResource;
  CHECK_FAILURE(textureBuffer->get_Resource(&dxgiResource));
  ComPtr<ID3D11Texture2D> sharedBuffer;
  CHECK_FAILURE(dxgiResource.As(&sharedBuffer));
  CHECK_FAILURE(deviceContext->CopyResource(sharedBuffer.Get(), inputTexture.Get()));

  // Notify the renderer for updated texture on the TextureBuffer.
  CHECK_FAILURE(webviewTextureStream->PresentBuffer(textureBuffer.Get(), timestamp));
}
```

# API Details
```
[v1_enum]
typedef enum COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND {
  /// CreateBuffer/Present and so on should return failed HRESULT if
  /// the texture stream is in the stopped state rather than using the
  /// error event. But there can be edge cases where the browser process
  /// knows the texture stream is in the stopped state and the host app
  /// process texture stream doesn't yet know that. Like the 10 second issue
  /// or if the script side has stopped the stream.
  COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED,
  /// The TextureBuffer already has been removed using CloseBuffer.

>   COREWEBVIEW2_TEXTURE_STREAM_ERROR_BUFFER_ERROR,

Reviewer: Rename buffer to texture per other comment about renaming buffer to texture.

      [out] EventRegistrationToken* token);
  /// Remove listener for StartRequest event.
  HRESULT remove_StartRequested(
      [in] EventRegistrationToken token);
  /// Listen to stop stream request once the stream started.
  /// It is called when user stop all streaming requests from
  /// the renderers (Javascript) or the host calls the Stop API. The renderer
  /// can stream again by calling the streaming request API.
  /// The renderer cleared all registered TextureBuffers before sending
  /// the stop request event so that the callback of the next start request
  /// should register the textures again.
  /// The event is triggered when all requests for given stream id closed
  /// by the Javascript, or the host's Stop API call.
  /// TextureBuffer related API calls after this event will return an error
  /// of HRESULT_FROM_WIN32(ERROR_INVALID_STATE).
  HRESULT add_Stopped(
      [in] ICoreWebView2StagingTextureStreamStoppedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for Stopped event.
  HRESULT remove_Stopped(
      [in] EventRegistrationToken token);

>   /// Creates TextureBuffer that will be referenced by the host and the browser.

Reviewer: Also update Buffer -> Texture and TextureBuffer -> Texture throughout documentation and parameter names

  /// By using the TextureBuffer mechanism, the host does not have to
  /// send respective texture to the renderer, instead it notifies it
  /// with internal TextureBuffer id, which is the identity of the TextureBuffer.
  /// The TextureBuffer is 2D texture, IDXGIResource, format and will be
  /// exposed through shared HANDLE or IUnknown type through ICoreWebView2StagingTexture.
  /// Whenever the host has new texture to write, it should ask
  /// reusable ICoreWebView2StagingTexture from the GetAvailableBuffer,
  /// which returns ICoreWebView2StagingTexture.
  /// If the GetAvailableBuffer returns an error, then the host calls the
  /// CreateBuffer to allocate new TextureBuffer.
  /// The API also registers created shared handle to the browser once it
  /// created the resource.

  /// Unit for width and height is texture unit (in texels).
  /// 'https://learn.microsoft.com/en-us/windows/win32/api/d3d11ns-d3d11-d3d11_texture2d_desc'

>   HRESULT CreateBuffer(

Reviewer: The method names and interface names are inconsistent. We should match between Buffer & Texture in the method names and interface names.

Proposed change:

TextureStream
 - Texture
 - CreateBuffer -> CreateTexture (rename)
 - PresentBuffer -> PresentTexture (rename)
 - *Buffer -> *Texture (rename)

    [in] UINT32 widthInTexels,
    [in] UINT32 heightInTexels,
    [out, retval] ICoreWebView2StagingTexture** buffer);
  /// Returns reuseable TextureBuffer for video frame rendering.
  /// Once the renderer finishes rendering of TextureBuffer's video frame, which
  /// was requested by Present, the renderer informs the host so that it can
  /// be reused. The host has to create new TextureBuffer with CreateBuffer
  /// if the API return an error HRESULT_FROM_WIN32(ERROR_NO_MORE_ITEMS).
  HRESULT GetAvailableBuffer([out, retval] ICoreWebView2StagingTexture** buffer);
  /// Removes TextureBuffer when the host removes the backed 2D texture.
  /// The host can save the resources by deleting 2D textures
  /// when it changes the frame sizes. The API will send a message
  /// to the browser where it will remove TextureBuffer.

>   HRESULT CloseBuffer([in] ICoreWebView2StagingTexture* buffer);

Reviewer: Add a note to spec to discuss with reviewers if this is better as
TextureStream.CloseTexture
or
Texture.Close

since it also involves the TextureStream managing sharing resources with the browser.

  /// Sets rendering image/resource through ICoreWebView2StagingTexture.
  /// The TextureBuffer must be retrieved from the GetAvailableBuffer or
  /// created via CreateBuffer.
  /// It is expected that hhe host writes new image/resource to the local
  /// shared 2D texture of the TextureBuffer (handle/resource).

  /// `timestampInMicroseconds` is video capture time with microseconds units.
  /// The value does not have to be exact captured time, but it should be
  /// increasing order because renderer (composition) ignores incoming
  /// video frame (texture) if its timestampInMicroseconds is equal or prior to
  /// the current compositing video frame. It also will be exposed to the
  /// JS with `VideoFrame::timestamp`.
  /// (https://docs.w3cub.com/dom/videoframe/timestamp.html).
  HRESULT PresentBuffer([in] ICoreWebView2StagingTexture* buffer,

>     [in] UINT64 timestampInMicroseconds);

Reviewer: Let's move the timestamp to the texture interface as a set/get property. The timestamp goes with the visual that's written to the texture so it makes sense that they are written together into one containing object.

  /// event.
  HRESULT SetD3DDevice([in] IUnknown* d3dDevice);
  /// Event handler for receiving texture by Javascript.
  /// `window.chrome.webview.registerTextureStream` call by Javascript will
  /// request sending video frame to the host where it will filter requested
  /// page's origin against allowed origins. If allowed, the Javascript will
  /// send a video frame (web texture), through MediaStreamTrack insertable APIs,
  /// MediaStreamTrackGenerator.
  /// https://www.w3.org/TR/mediacapture-transform/.
  /// WebTextureReceived event will be called only when it receives
  /// a web texture. There is no start event for receiving web texture.
  HRESULT add_WebTextureReceived(
      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for WebTextureReceived event.
  HRESULT remove_WebTextureReceived([in] EventRegistrationToken token);
  /// Event handler for stopping of the receiving texture stream.
  /// It is expected that the host releases any holding handle/resource from
  /// the WebTexture before an event handler returns.

>   HRESULT add_WebTextureStreamStopped(

Reviewer: We need to update this documentation to say what JavaScript APIs or situations result in this event getting raised. Its only during the call to unregister.

>       [in] ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler* eventHandler,

Reviewer: We need to update the documentation to say that a later call to register may start the stream again, and you will know when you start receiving WebTextureReceived events.

      [out] EventRegistrationToken* token);
  /// Remove listener for WebTextureStreamStopped event.
  HRESULT remove_WebTextureStreamStopped([in] EventRegistrationToken token);
}
/// TextureBuffer that the host writes to so that the Renderer
/// will render on it.
[uuid(0836f09c-34bd-47bf-914a-99fb56ae2d07), object, pointer_default(unique)]
interface ICoreWebView2StagingTexture : IUnknown {

>     /// Returns shared Windows NT handle. The caller expected to open it with

Reviewer: ```suggestion
    /// Returns Windows NT handle to shared memory containing the texture. The caller is expected to open it with
```

    /// ID3D11Device1::OpenSharedResource1 and writes the incoming texture to it.

>     [propget] HRESULT Handle([out, retval] HANDLE* handle);

Reviewer: Update documentation to say that the caller should not close the handle. The texture object owns closing it. The underlying shared memory may be reused by WebView2.

interface ICoreWebView2StagingTextureStreamStoppedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event. There are no event args and the args
  /// parameter will be null.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);
}
/// This is the callback for texture stream rendering error.
[uuid(52cb8898-c711-401a-8f97-3646831ba72d), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamErrorReceivedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] ICoreWebView2StagingTextureStreamErrorReceivedEventArgs* args);
}
/// This is the event args interface for texture stream error callback.
[uuid(0e1730c1-03df-4ad2-b847-be4d63adf700), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamErrorReceivedEventArgs : IUnknown {
  /// Error kind.
  [propget] HRESULT Kind([out, retval]
      COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND* value);

>   /// TextureBuffer that the error is associated with.

Reviewer: Documentation should note that for some kinds of errors there is no corresponding texture and the Texture property will return nullptr.

>   HRESULT GetBuffer([out, retval] ICoreWebView2StagingTexture** buffer);

Reviewer: Should be a property to get the texture:

`[propget] HRESULT Texture(([out, retval] ICoreWebView2StagingTexture** buffer);`

  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingEnvironment * sender,
      [in] IUnknown* args);
}
/// This is the callback for web texture.
[uuid(9ea4228c-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs* args);
}

/// This is the event args interface for web texture.
[uuid(a4c2fa3a-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs : IUnknown {
  /// Return ICoreWebView2StagingWebTexture object.
  /// The call does not create new ICoreWebView2StagingWebTexture object, instead
  /// returns the same object.

  /// The TextureBuffer handle will be reused when ICoreWebView2StagingWebTexture
  /// object is released. So, the host should not refer handle or resource of
  /// the ICoreWebView2StagingWebTexture after its release.

>   HRESULT GetWebTexture([out, retval] ICoreWebView2StagingWebTexture** value);

Reviewer: This can be just a property not a method:

`[propget] HRESULT WebTexture([out, retval] ICoreWebView2StagingWebTexture** value);`

}

/// This is the callback for web texture stop.
[uuid(77eb4638-2f05-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);
}

/// Received TextureBuffer that the renderer writes to so that the host
/// will read on it.
[uuid(b94265ae-4c1e-11ed-bdc3-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingWebTexture : IUnknown {
  /// TextureBuffer handle. The handle's lifetime is owned by the
  /// ICoreWebView2StagingTextureStream object so the host must not close it.
  /// The same handle value will be used for same TextureBuffer so the host can use
  /// handle value as a unique TextureBuffer key.
  /// If the host opens its own resources by handle, then it is suggested
  /// that the host removes those resources when the handle's texture size
  /// is changed because the browser also removed previously allocated different
  /// sized buffers when image size is changed.

>   [propget] HRESULT Handle([out, retval] HANDLE* handle);

Reviewer: Just make sure to update these docs the same as changes to the other Texture interface's docs.

Author: use value for property.

  /// TextureBuffer resource.
  /// The same resource value will be used for same TextureBuffer so the host can use
  /// resource value as a unique TextureBuffer key.
  /// ICoreWebView2StagingTextureStream object has a reference of the resource
  /// so ICoreWebView2StagingWebTexture holds same resource object for
  /// the same TextureBuffer.

>   [propget] HRESULT Resource([out, retval] IUnknown** resource);

Reviewer: For all propget/ propput methods, the parameter is supposed to be named 'value'

```suggestion
  [propget] HRESULT Resource([out, retval] IUnknown** value);
```

The API will use the shared GPU texture buffer so that it can minimize the overall cost with
regards to frame copy.

# Examples

## Javascript

This is Javascript code common to both of the following samples:

```js
// getTextureStream sample.

// Developer can stream from the host and its returned object is MediaStream
// that is same what getUserMedia returns. Technically, developer can do all
// what MediaStream and its video MediaStreamTrack provide.

// Scenario: User clicks the button and show stream from the host via video element.
document.querySelector('#showVideo').addEventListener('click',
  e => getStreamFromTheHost(e));
async function getStreamFromTheHost(e) {
  try {
    // Request stream to the host with unique stream id.
    // getTextureStream W3C standard MediaStream object that has only video
    // MediaStreamTrack.
    const stream = await window.chrome.webview.getTextureStream('webview2-abcd1234');
    // The MediaStream object is returned and it gets video MediaStreamTrack element from it.
    const videoTracks = stream.getVideoTracks();
    const videoTrack = videoTracks[0];
    window.videoTrack = videoTrack;
    // Show the video via Video Element.

>     let videoId = document.createElement('video');

Reviewer: videoId should probably be called videoElement

  EventRegistrationToken stopped_token;
  CHECK_FAILURE(webviewTextureStream->add_WebTextureStreamStopped(Callback<ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler>(
    [&](ICoreWebView2StagingTextureStream* sender, IUnknown* args) {

      return S_OK;
    }).Get(), &stopped_token));
}  // CreateTextureStream

HRESULT CallWebView2API(bool createBuffer,
                        UINT32 width,
                        UINT32 height,
                        bool requestAvailableBufer,
                        bool sendTexture,
                        ICoreWebView2StagingTexture* send_texture,
                        UINT64 timestamp,
                        ICoreWebView2StagingTexture** texture) {
  if (createBuffer) {
    // Create shared buffer.
    return webviewTextureStream->CreateSharedBuffer(width, height, &texture);
  }

  if (requestAvailableBufer) {
    return webviewTextureStream->GetAvailableBuffer(&texture);
  }

  if (sendTexture) {
    // Notify the renderer for updated texture on the shared buffer.
    webviewTextureStream->SetBuffer(send_texture, timestamp);
    webviewTextureStream->Present();
  }
}
```

> # API Details

Reviewer: Missing sample code and MIDL3 for WinRT/C# APIs. Is that intentional? If so, it needs to be mentioned in the document at the top somewhere about why

Author: yes, it is intentional.

/// The proposed APIs have dependency on the DirectX and its internal attributes such as
/// adapter LUID so it supports Win32/C++ and C++/WinRT APIs at this time.

Author: Then, I will remove C++/WinRT as developers can use the APIs though like UWP DirectX app in the official Microsoft UWP samples (https://learn.microsoft.com/en-us/windows/uwp/gaming/tutorial--create-your-first-uwp-directx-game, https://github.com/microsoft/Windows-universal-samples/tree/main/Samples/Simple3DGameDX)

}
/// This is the interface that handles texture streaming.
[uuid(afca8431-633f-4528-abfe-7fc3bedd8962), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStream : IUnknown {
  /// Get the stream ID of the object that is used when calling CreateTextureStream.
  /// The caller must free the returned string with CoTaskMemFree. See
  /// [API Conventions](/microsoft-edge/webview2/concepts/win32-api-conventions#strings).
  // MSOWNERS: TBD (wv2core@microsoft.com)
  [propget] HRESULT StreamId([out, retval] LPWSTR* id);
  /// Adds an allowed url origin for the given stream id. The stream requests
  /// could be made from any frame, including iframes, but these origins
  /// must be registered first in order for the request to succeed.
  /// The added filter will be persistent until
  /// ICoreWebView2StagingTextureStream is destroyed or
  /// RemoveAllowedOrigin is called.
  /// The renderer does not support wildcard so it will compare
  /// literal string input to the requesting frame origin. So, the input string
  /// should have a scheme like https://.
  /// For example, https://www.valid-host.com, http://www.valid-host.com are

>   /// valid origins but  www.valid-host.com, or *.valid-host.com. are not

Author: yes, it return an error of E_INVALIDARG. will add a comment of error info.

  /// valid origins.
  /// getTextureStream() will fail unless the requesting frame's origin URL is
  /// added to the request filter.
  HRESULT AddAllowedOrigin([in] LPCWSTR origin);

>   /// Remove added origin, which was added by AddAllowedOrigin.

Reviewer: Need to describe when the remove is applied. If we're currently streaming textures to an origin and in the middle of that we call remove does that origin get cut off right away? Or it only blocks future startRequested calls? I think either is fine, but we need to document it.

Author: /// The allowed or disallowed origins will take effect only when Javascript
  /// request a streaming. So, once the streaming started, it does not stop
  /// streaming.

  /// Remove added origin, which was added by AddRequestedFilter.
  HRESULT RemoveRequestedFilter([in] LPCWSTR origin);
  /// Listens for stream requests from the Javascript's getTextureStream call
  /// for the given stream id. It is called for the first request only, the

>   /// subsequent requests of same stream id will not be called.

Reviewer: Would be good to elaborate here. I assume its like the StartRequested event won't be raised if we're already waiting the 10s for an already pending StartRequested and that it won't be raised if the StartRequested is completed and the we're running. But if we stop after that or the 10s expires that StartRequested will be raised again?

>   /// The request is regarded as success only when the host provides the stream,

Author: If we look at in terms of state, yes it has only Start and Stop state.  Javascript side, it will follow the existing MediaStream lifetime, which is from w3c standard. https://www.w3.org/TR/mediacapture-streams/ AS the return object is MediaStream on JS side.

  /// The Texture stream becomes 'Started' state once it starts sending a texture
  /// and becomes 'Stopped' state on the moment when it receives StopRequested.

>   /// The request is regarded as success only when the host provides the stream,

Author: "What happens when native calls Present but JS never called Start? " 
It is never started. (Present call (within 10s) after StartRequested is valid).

"Is the texture stream 'started' and if JS tries to call 'Start' it will just immediately start receiving what the texture stream has already been sending"
No with above reason.

  /// Present API call, within 10s after being requested.
  HRESULT add_StartRequested(
      [in] ICoreWebView2StagingTextureStreamStartRequestedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for start stream request.
  HRESULT remove_StartRequested(
      [in] EventRegistrationToken token);
  /// Listen to stop stream request once the stream started.
  /// It is called when user stop all streaming requests from
  /// the renderers (Javascript) or the host calls the Stop API. The renderer
  /// can stream again by calling the streaming request API.
  /// The renderer cleared all registered buffers before sending
  /// the stop request event so that the callback of the next start request
  /// should register the textures again.
  /// The event is triggered when all requests for given stream id closed
  /// by the Javascript, or the host's Stop API call.

>   HRESULT add_StopRequested(

Author: (I'm not sure i might already replied on other comment. somehow view doesn't work well).

The host is a streaming provider that means it is the one that decide whether it will start or stop streaming. The browser in this case is just client who can request of it.  

client: Start request means, I'm ready to accept the texture, please send me texture.
host: ok, I will send a texture (Started state) or no I don't want to (Stopped state).

client: Stop request means, I don't want to listen to stream from you so I've already clean any resources for it. you don't have to send me a texture.
host: ok, I will stop sending you (Stopped state), or hmm I want to keep sending you texture (Stopped state) although no one listen to.

It looks like we are in two different worlds so it is easier to think Start/State is from the host side perspective as the API is host face one.

      [in] ICoreWebView2StagingTextureStreamStopRequestedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for stop stream request.
  HRESULT remove_StopRequested(
      [in] EventRegistrationToken token);

>   /// Creates shared buffer that will be referenced by the host and the browser.

Author: TextureBuffer sounds good to me, I will change comments across the doc.

  /// By using the shared buffer mechanism, the host does not have to
  /// send respective texture to the renderer, instead it notifies it
  /// with internal buffer id, which is the identity of the shared buffer.
  /// The shared buffer is 2D texture, IDXGIResource, format and will be
  /// exposed through shared HANDLE or IUnknown type through ICoreWebView2StagingTexture.
  /// Whenever the host has new texture to write, it should ask
  /// reusable ICoreWebView2StagingTexture from the GetAvailableBuffer,
  /// which returns ICoreWebView2StagingTexture.
  /// If the GetAvailableBuffer returns an error, then the host calls the
  /// CreateBuffer to allocate new shared buffer.
  /// The API also registers created shared handle to the browser once it
  /// created the resource.
  HRESULT CreateBuffer(

>     [in] UINT32 width,

Author: It is a texture unit (texel).

Added comment.
"  /// Unit for width and height is texture unit (in texels).
  /// 'https://learn.microsoft.com/en-us/windows/win32/api/d3d11ns-d3d11-d3d11_texture2d_desc'''"

>     [in] UINT32 width,

Author: yes, it can't be. and it has a maximum limit as well, which can be different by its device API feature level. I've added a link of https://learn.microsoft.com/en-us/windows/win32/api/d3d11/ns-d3d11-d3d11_texture2d_desc to the comment that can be much easier for developers rather than copy and paste the unit section from the link.

>     [in] UINT32 width,

Author: done,

    [in] UINT32 height,
    [out, retval] ICoreWebView2StagingTexture** buffer);
  /// GetAvailableBuffer can be called on any thread like SetBuffer.

>   HRESULT GetAvailableBuffer([out, retval] ICoreWebView2StagingTexture** buffer);

Author: /// Once the renderer finishes rendering of image in the TextureBuffer, then
  /// the host can reuse the buffer again. GetAvailableBuffer API will return
  /// any reusable TextureBuffer if there is any.

>   HRESULT GetAvailableBuffer([out, retval] ICoreWebView2StagingTexture** buffer);

Author: We use shared handle mechanism between the host and the browser. The textureBuffer that is created from the host will be sent to the browser with internal buffer id. The browser will open the passed handle and render it.  Once the browser render it, it sends a message to the host so the host could reuse the textureBuffer, which is a return of GetAvailableBuffer.   The host write its video image to this available buffer, and notify the browser with the buffer id where it returns the handle that is associated with this buffer id. 

This mechanism is already existing for handling camera capture where TextureStream use as much as possible of it.

  /// Remove texture buffer when the host removes the backed 2D texture.
  /// The host can save the existing resources by deleting 2D textures
  /// when it changes the frame sizes.

>   HRESULT RemoveBuffer([in] ICoreWebView2StagingTexture* buffer);

Author: Actually, texture is a handle/resource. texture is more graphics concept and handle/resource is API representation that developer can access the texture; it follows DirectX naming. 

It disconnects given TextureBuffer (buffer_id) from the streaming.

  /// Indicates that the buffer is ready to present.
  /// The buffer must be retrieved from the GetAvailableBuffer.
  /// The host writes new texture to the local shared 2D texture of
  /// the buffer id, which is created via CreateBuffer.
  /// SetBuffer API can be called in any thread.
  HRESULT SetBuffer([in] ICoreWebView2StagingTexture* buffer,

>     [in] ULONGLONG timestamp);

Author: /// timestamp is video capture time with unit of 100-nanosecond units.
  /// The value doesn't have to be exact captured time, but it should be
  /// increasing order, the next Present's TextureBuffer should have later
  /// time.

  /// Render texture that is current set ICoreWebView2StagingTexture.

>   HRESULT Present();

Reviewer: Is there a reason why SetBuffer and Present are separate? Present renders the texture that was last set via SetBuffer (right?). Should they be combined and Present should take the SetBuffer parameters? Or is there a reason to call SetBuffer and then wait to call Present?

Author: It is feedback from the Edge graphics side code reviewer; Present is an API that DirectX to use for commit rendering, and SetBuffer is kind of Swap to front buffer. It is said that it is familiar pattern with the graphics developer.

Reviewer: > I've addressed comments, please reactivate anything it isn't clear to you. Thanks a lot.

Some of my comments are still unresolved. For example:
 * https://github.com/MicrosoftEdge/WebView2Feedback/pull/2743/files/bc5d2674edecfdf303bdf8d28308a44a61eee0be#r1020474503
 * https://github.com/MicrosoftEdge/WebView2Feedback/pull/2743/files/bc5d2674edecfdf303bdf8d28308a44a61eee0be#r1020472689
 * https://github.com/MicrosoftEdge/WebView2Feedback/pull/2743/files/bc5d2674edecfdf303bdf8d28308a44a61eee0be#r1020472927

Is that intentional?

Reviewer: That's great. We'll just need to ensure that the documentation is clear.

  /// Stop streaming of the current stream id.
  /// API calls of Present, CreateBuffer will fail after this
  /// with an error of COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED.
  /// The Javascript can restart the stream with getTextureStream.
  HRESULT Stop();
  /// Event handler for those that occur at the Renderer side, the example
  /// are CreateBuffer, Present, or Stop.
  HRESULT add_TextureError(
      [in] ICoreWebView2StagingTextureStreamTextureErrorEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for texture error event.
  HRESULT remove_TextureError([in] EventRegistrationToken token);
  /// Updates d3d Device when it is updated by RenderAdapterLUIDChanged
  /// event.
  HRESULT UpdateD3DDevice([in] IUnknown* d3dDevice);
  /// Event handler for receiving texture by Javascript.
  /// There is no Start event for web texture. Whenever texture are sent

>   /// from the Javascript, the event is triggered.

Reviewer: It would be good to reference the exact DOM JS APIs used to send textures that raise this event

Author: /// Event handler for receiving texture by Javascript.
  /// `window.chrome.webview.registerTextureStream` call by Javascript will
  /// request sending video frame to the browser where it will allow or
  /// disallow based on allowed origins, `AddWebTextureAllowedOrigin`.
  /// Once it is allowed, the Javascript will send a video frame through
  /// MediaStreamTrack insertable APIs, MediaStreamTrackGenerator.
  /// https://www.w3.org/TR/mediacapture-transform/

  HRESULT add_WebTextureReceived(
      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);

>   HRESULT remove_WebTextureReceived([in] EventRegistrationToken token);

Reviewer: For the remove_* methods just a simple reference documentation like should be above each. This one and the next one.

  /// Remove listener for the {event name} event.

>   /// Event handler for stopping of the receiving texture stream.

Reviewer: More documentation required here. When is this raised? In relation to something happening in JS - a particular JS DOM API?

Author: /// Remove listener for receiving texture stream.

  HRESULT add_WebTextureStreamStopped(
      [in] ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  HRESULT remove_WebTextureStreamStopped([in] EventRegistrationToken token);

  /// Adds an allowed url origin for the given stream id for web texture
  /// operation.  Javascript can send texture stream to the host only when
  /// the origin it runs are allowed by the host.

>   HRESULT AddWebTextureAllowedOrigin([in] LPCWSTR origin);

Reviewer: Why do we have to have a separate allow list for this? Can we just use the one AddAllowedOrigin? 

If we do need it separate how about modifying AddAllowedOrigin to take a permission parameter. I guess its not possible to have an origin that's not allowed to receive textures but is allowed to send them. 
So you can do 

AddAllowedOrigin("https://example.com", true) where the second param is a bool if they're allowed to send textures back.

Author: It is requested API from the Teams because of their security review board. They prefer independent APIs for respective get and register texture stream case, which is what Electron based solution has it.

Reviewer: We can still have the separate features with separate permissions, but one method AddAllowedOrigin where you specify the permission per origin. This lets us ensure that you can't get into a weird state of an origin allowed to send textures but not allowed to receive textures. It should also make it easier to read and audit code to see what origins get what permissions.

Author: It might cause more confusion to the developer whether they need to add false or true when calling RemoveAllowedOrigin? for example, remove bidi only, or remove both of them, (potentially normal direction only). 

For normal direction only, there was a request to allow registerTextureStream without getTextureStream a while ago just in case of future use from the Teams, but we rejected it at that time as it isn't supported in the current Electron app.  I'm afraid though if we have single API for both directions, we might have to create API again if the request becomes a product request in future.

We expect that most common scenario is normal direction only, only certain scenario needs bidi (registerTextureStream).

However, I'm open to change though, please let me know what you think.

Author: I've talked with Teams team about this and they are fine with single API, I will combine these two into single AddAlloedOrigin.

Author: Change as suggested.

  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);
}
[uuid(431721e0-0f18-4d7b-bd4d-e5b1522bb110), object, pointer_default(unique)]
interface ICoreWebView2StagingRenderAdapterLUIDChangedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2Staging3 * sender,
      [in] IUnknown* args);
}
/// This is the callback for web texture.
[uuid(9ea4228c-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs* args);
}

/// This is the event args interface for web texture.
[uuid(a4c2fa3a-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs : IUnknown {

>   // Return ICoreWebView2StagingWebTexture object.

Reviewer: Should be /// to produce our reference documentation

```suggestion
  /// Return ICoreWebView2StagingWebTexture object.

```

  [propget] HRESULT Handle([out, retval] HANDLE* handle);

  /// Texture buffer resource.
  /// The same resource value will be used for same buffer so the host can use
  /// resource value as a unique buffer key.
  /// ICoreWebView2StagingTextureStream object has a reference of the resource
  /// so ICoreWebView2StagingWebTexture holds same resource object for
  /// the same buffer.
  [propget] HRESULT Resource([out, retval] IUnknown** resource);

  /// It is timestamp of the web texture. Javascript can set this value
  /// with any value, but it is suggested to use same value of its original
  /// video frame that is a value of SetBuffer so that the host is able to
  /// tell the receiving texture delta.
  [propget] HRESULT Timestamp([out, retval] ULONGLONG* value);
}

/// This is the callback for web texture stop.
[uuid(77eb4638-2f05-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);

> }/// This is the callback for web texture.

Reviewer: All of the interfaces after this line are duplicates of the above

Author: I'm sorry for confusion.

[uuid(9ea4228c-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs* args);
}

/// This is the event args interface for web texture.
[uuid(a4c2fa3a-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs : IUnknown {

>   // Return ICoreWebView2StagingWebTexture object.

Reviewer: Use /// for generating reference documentation.
```suggestion
  /// Return ICoreWebView2StagingWebTexture object.
```

TextureStream
===============================================================================================

# Background

> Many native apps use a native engine for real-time communication scenarios, which include video capture, networking and video rendering.  However, often, these apps still use WebView or Electron for UI rendering. The separation between real-time video rendering and UI rendering prevents apps from rendering real-time video inside the web contents. This forces apps to render the real-time video on top of the web contents, which is limiting. Rendering video on top constrains the user experience and it may also cause performance problems. We can ask the native apps to use web renderer for video handling because web standard already provides these features through WebRTC APIs. The end developers, however, prefer to use their existing engine such as capturing and composition, meanwhile using WebRTC API for rendering.

Reviewer: Including the lines I recommend the text changes below

  /// exposed through shared HANDLE or IUnknown type through ICoreWebView2StagingTexture.
  /// Whenever the host has new texture to write, it should ask
  /// reusable ICoreWebView2StagingTexture from the GetAvailableTexture,
  /// which returns ICoreWebView2StagingTexture.
  /// If the GetAvailableTexture returns an error, then the host calls the
  /// CreateTexture to allocate new texture.
  /// The API also registers created shared handle to the browser once it
  /// created the resource.

  /// Unit for width and height is texture unit (in texels).
  /// 'https://learn.microsoft.com/en-us/windows/win32/api/d3d11ns-d3d11-d3d11_texture2d_desc'
  HRESULT CreateTexture(
    [in] UINT32 widthInTexels,
    [in] UINT32 heightInTexels,
    [out, retval] ICoreWebView2StagingTexture** texture);
  /// Returns reuseable texture for video frame rendering.
  /// Once the renderer finishes rendering of texture's video frame, which
  /// was requested by Present, the renderer informs the host so that it can
  /// be reused. The host has to create new texture with CreateTexture
  /// if the API return an error HRESULT_FROM_WIN32(ERROR_NO_MORE_ITEMS).

>   HRESULT GetAvailableTexture([out, retval] ICoreWebView2StagingTexture** texture);

Author: There is usage case where the native creates many static numbers of textures by calling CreateTextures, and recycles them without calling GetAvailableTexture yet all.   It is efficient way as long as the machine has enough resources.  The Teams also use this way, and use GetAvailable only when the machine has low resources.

If we create GetOrCreate API, then they can't use static textures method.

  /// Removes texture when the host removes the backed 2D texture.
  /// The host can save the resources by deleting 2D textures
  /// when it changes the frame sizes. The API will send a message
  /// to the browser where it will remove texture.
  HRESULT CloseTexture([in] ICoreWebView2StagingTexture* texture);

>   /// Sets rendering image/resource through ICoreWebView2StagingTexture.

Author: yes, E_INVALIDARG is correct.

  /// The texture must be retrieved from the GetAvailableTexture or
  /// created via CreateTexture.
  /// It is expected that hhe host writes new image/resource to the local
  /// shared 2D texture of the texture (handle/resource).
  HRESULT PresentTexture([in] ICoreWebView2StagingTexture* texture)

>   /// Stop streaming of the current stream id.

Reviewer: COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED is from the enum not an HRESULT right? We should document the HRESULT that gets returned if you call one of the methods while the texture stream is stopped.

```suggestion
  /// Stops this texture stream from streaming and moves it into the stopped state. When moving to the stopped state the `ICoreWebView2TextureStream Stopped` event will be raised and in script the `MediaStreamTrack ended` event will be raised. Once stopped, script may again call `Webview.getTextureStream` moving the texture stream back to the start requested state. See the `StartRequested` event for details.
  /// Once stopped, calls to CreateTexture, GetAvailableTexture, PresentTexture, and CloseTexture will fail with COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED. See those methods for details.
  /// The `Stop` method is implicitly called when the texture stream object is destroyed.
```

  /// The Javascript will receive `MediaStreamTrack::ended` event when the API
  /// is called.
  /// The Javascript can restart the stream with getTextureStream.
  /// The API call will release any internal resources on both of WebView2 host
  /// and the browser processes.
  /// API calls of Present, CreateTexture will fail after this
  /// with an error of COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED.
  /// The Stop API will be called implicitly when ICoreWebView2StagingTextureStream
  /// object is destroyed.
  HRESULT Stop();

>   /// Event handler for those that occur at the Renderer side, the example

Reviewer: ```suggestion
  /// The `ErrorReceived` event is raised when an error with this texture stream occurs asynchronously.
```

  /// are CreateTexture, Present, or Stop.
  HRESULT add_ErrorReceived(
      [in] ICoreWebView2StagingTextureStreamErrorReceivedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for ErrorReceived event.
  HRESULT remove_ErrorReceived([in] EventRegistrationToken token);

>   /// Updates d3d Device when it is updated by RenderAdapterLUIDChanged

Reviewer: ```suggestion
  /// Set the D3D device this texture stream should use for creating shared texture resources. When the RenderAdapterLUIDChanged event is raised you should create a new D3D device using the RenderAdapterLUID property and call SetD3DDevice with the new D3D device. See the `CreateTextureStream` `d3dDevice` parameter for more details.
```

  /// The same handle value will be used for same texture so the host can use
  /// handle value as a unique texture key.
  /// If the host opens its own resources by handle, then it is suggested
  /// that the host removes those resources when the handle's texture size
  /// is changed because the browser also removed previously allocated different
  /// sized textures when image size is changed.
  [propget] HRESULT Handle([out, retval] HANDLE* value);

  /// Direct2D texture resource.
  /// The same resource value will be used for same texture so the host can use
  /// resource value as a unique texture key.
  /// ICoreWebView2StagingTextureStream object has a reference of the resource
  /// so ICoreWebView2StagingWebTexture holds same resource object for
  /// the same texture.
  [propget] HRESULT Resource([out, retval] IUnknown** value);

  /// It is timestamp of the web texture. Javascript can set this value
  /// with any value, but it is suggested to use same value of its original
  /// video frame that is a value of PresentTexture so that the host is able to
  /// tell the receiving texture delta.
  [propget] HRESULT Timestamp([out, retval] UINT64* value);
}

```

> ---ts

Reviewer: This should be ` ` ` not - - -. I did --- previously just because I wanted the comment to show up well in the PR issue box

  ComPtr<D3D11Device> d3d_device = MyCreateD3DDevice(luid);

  // Register unique texture stream that the host can provide.
  ComPtr<ICoreWebView2TextureStream> webviewTextureStream;
  CHECK_FAILURE(environment->CreateTextureStream(L"webview2-abcd1234",
      d3d_device.Get(),  &webviewTextureStream));

  // Register the Origin URI that the target renderer could stream of the registered
  // stream id. The request from not registered origin will fail to stream.

  // `true` boolean value will add allowed origin for registerTextureStream as well.
  CHECK_FAILURE(webviewTextureStream->AddAllowedOrigin(
      L"https://edge-webscratch"), true);

  // Listen to Start request. The host will setup system video streaming and
  // start sending the texture.
  EventRegistrationToken start_token;
  CHECK_FAILURE(webviewTextureStream->add_StartRequested(Callback<ICoreWebView2StagingTextureStreamStartRequestedEventHandler>(
    [hWnd](ICoreWebView2StagingTextureStream* webview,
          IUnknown* eventArgs) -> HRESULT {
      // Capture video stream by using native API, for example,
      // Media Foundation on Windows.
      StartMediaFoundationCapture(hWnd);
      return S_OK;

>     }).Get(), &start_token));ffer

Reviewer 2: ```suggestion
    }).Get(), &start_token));
```

  /// You should write your video frame data to the `ICoreWebView2Texture`
  /// before calling this method.

  /// After this method completes WebView2 will take some time asynchronously
  /// to send the texture to the WebView2 processes to be added to the video stream.
  /// Do not close or otherwise change the provided `ICoreWebView2Texture` after
  /// calling this method. Doing so may result in the texture not being added to
  /// the texture stream and the `ErrorReceived` event may be raised.
  HRESULT PresentTexture([in] ICoreWebView2StagingTexture* texture)

  /// Stops this texture stream from streaming and moves it into the stopped state.
  /// When moving to the stopped state the `ICoreWebView2TextureStream Stopped`
  /// event will be raised and in script the `MediaStreamTrack ended` event will
  /// be raised. Once stopped, script may again call `Webview.getTextureStream`
  /// moving the texture stream back to the start requested state.
  /// See the `StartRequested` event for details.

  /// Once stopped, calls to CreateTexture, GetAvailableTexture,
  /// PresentTexture, and CloseTexture will fail with

>   /// COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED.

Reviewer: This seems wrong. This is an enum value not an HRESULT. What will the methods return when in the stopped state?

Author: It will return HRESULT_FROM_WIN32(ERROR_INVALID_STATE).

  /// See those methods for details.

  /// The `Stop` method is implicitly called when the texture stream object is
  /// destroyed.
  HRESULT Stop();
  /// The `ErrorReceived` event is raised when an error with this texture
  /// stream occurs asynchronously.
  HRESULT add_ErrorReceived(
      [in] ICoreWebView2StagingTextureStreamErrorReceivedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for ErrorReceived event.
  HRESULT remove_ErrorReceived([in] EventRegistrationToken token);
  /// Set the D3D device this texture stream should use for creating shared
  /// texture resources. When the RenderAdapterLUIDChanged event is raised you
  /// should create a new D3D device using the RenderAdapterLUID property and
  /// call SetD3DDevice with the new D3D device.
  /// See the `CreateTextureStream` `d3dDevice` parameter for more details.
  HRESULT SetD3DDevice([in] IUnknown* d3dDevice);

>   /// Event handler for receiving texture by Javascript.

Reviewer: ```suggestion
    /// The WebTextureReceived event is raised when script sends a video frame to this texture stream. Allowed script will call `chrome.webview.registerTextureStream` to register a MediaStream with a specified texture stream. Video frames added to that MediaStream will be raised in the WebTextureReceived event. See `registerTextureStream` for details. Script is allowed to call registerTextureStream if it is from an HTML document with an origin allowed via `ICoreWebView2TextureStream::AddAllowedOrigin` with the `alllowWebTexture` parameter set. See `AddAllowedOrigin` for details.
```

  /// `window.chrome.webview.registerTextureStream` call by Javascript will
  /// request sending video frame to the host where it will filter requested
  /// page's origin against allowed origins. If allowed, the Javascript will
  /// send a video frame (web texture), through MediaStreamTrack insertable APIs,
  /// MediaStreamTrackGenerator.
  /// https://www.w3.org/TR/mediacapture-transform/.
  /// WebTextureReceived event will be called only when it receives
  /// a web texture. There is no start event for receiving web texture.
  HRESULT add_WebTextureReceived(
      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for WebTextureReceived event.
  HRESULT remove_WebTextureReceived([in] EventRegistrationToken token);

>   /// Event handler for stopping of the receiving texture stream.

Reviewer: Do I understand your note "It is expected that the host releases any holding handle/resource from the WebTexture before an event handler returns" that you will close the web textures and the underlying shared memory resources for them after the event handler completes? If so, you need to add a GetDeferral method (see NewWindowRequestedEventArgs.GetDeferral for an example) to the event args. This lets the host app use the deferral to say when the event handler completes asynchronously in case any of the web texture cleanup work that the host needs to do requires running async methods or otherwise requires async work.

```suggestion
  /// The WebTextureStreamStopped event is raised when script unregisters its MediaStream from this texture stream. Script that has previously called `chrome.webview.registerTextureStream`, can call `chrome.webview.unregisterTextureStream` which will raise this event and then close associated ICoreWebView2WebTexture objects after the event handler completes. You should ensure that you release any references to associated ICoreWebView2WebTexture objects and their underlying resources before the event handler completes.
  /// Once stopped, script may start again by calling `chrome.webview.registerTextureStream` and sending more frames. In this case the `ICoreWebView2TextureStream WebTextureReceived` event will be raised again.
```

Author: WetTextureStopped event is a notification that the browser already released all shared resources. It is expected that the host cleans any resources the host created for supporting the received web texture.

slightly rephrase the first one.
"/// The WebTextureStreamStopped event is raised when script unregisters its
  /// MediaStream from this texture stream. Script that has previously called
  /// `chrome.webview.registerTextureStream`, can call `chrome.webview.
  /// unregisterTextureStream` which will raise this event and then close
  /// associated ICoreWebView2WebTexture objects in the browser side. You should
  /// ensure that you release any references to associated
  /// ICoreWebView2WebTexture objects and their underlying resources."

  /// It is expected that the host releases any holding handle/resource from
  /// the WebTexture before an event handler returns. The event can be raised
  /// when the JS calls `window.chrome.webview.unregisterTextureStream`.
  /// JS can restart sending the stream with `window.chrome.webview.registerTextureStream`
  /// after stream is stopped. If the stream is started again, `add_WebTextureReceived`
  /// will be called.
  HRESULT add_WebTextureStreamStopped(
      [in] ICoreWebView2StagingTextureStreamWebTextureStreamStoppedEventHandler* eventHandler,
      [out] EventRegistrationToken* token);
  /// Remove listener for WebTextureStreamStopped event.
  HRESULT remove_WebTextureStreamStopped([in] EventRegistrationToken token);
}

> /// texture that the host writes to so that the Renderer

Reviewer: ```suggestion
/// A shared texture object used by ICoreWebView2TextureStream to represent a video frame. You can obtain them via `ICoreWebView2TextureStream::CreateTexture`, write to the underlying shared memory, and then have it rendered to a video stream with `ICoreWebView2TextureStream::PresentTexture`. See those methods for details.
```

/// will render on it.
[uuid(0836f09c-34bd-47bf-914a-99fb56ae2d07), object, pointer_default(unique)]
interface ICoreWebView2StagingTexture : IUnknown {

>     /// Returns Windows NT handle to shared memory containing the texture.

Reviewer: What do we say about closing the handle? The caller shouldn't call CloseHandle or anything like that right? The underlying memory is owned by WebView2 and WebView2 will close it.

```suggestion
    /// A handle to OS shared memory containing the texture. You can open it with `ID3D11Device1::OpenSharedResource1` and write your texture data to it. Do not close it yourself. The underlying memory will be closed by WebView2. Do not change the memory after calling `ICoreWebView2TextureStream::PresentTexture` or you the frame may not be rendered and the `ICoreWebView2TextureStream ErrorReceived` event will be raised.
```

Author: thanks, slightly rephrase.

    /// A handle to OS shared memory containing the texture. You can open it
    /// with `ID3D11Device1::OpenSharedResource1` and write your texture data
    /// to it. Do not close it yourself. The underlying texture will be closed
    /// by WebView2. Do not change the texture after calling
    /// `ICoreWebView2TextureStream::PresentTexture` before you can retrieve it
    /// again with `GetAvailableTexture`, or you the frame may not be
    /// rendered and the `ICoreWebView2TextureStream ErrorReceived` event will
    /// be raised.

    /// The caller expected to open it with ID3D11Device1::OpenSharedResource1
    /// and writes the incoming texture to it.
    [propget] HRESULT Handle([out, retval] HANDLE* value);

>     /// Returns IUnknown type that could be query interface to IDXGIResource.

Reviewer: ```suggestion
    /// The texture as an `IUnknown` that supports the `IDXGIResource` interface. You can use the `IDXGIResource` interface to write your texture data to the texture. Do not change the texture after calling `ICoreWebView2TextureStream::PresentTexture` or you the frame may not be rendered and the `ICoreWebView2TextureStream ErrorReceived` event will be raised.
```

    /// The caller can write incoming texture to it.
    [propget] HRESULT Resource([out, retval] IUnknown** value);

>     /// Sets timestamp of presenting texture.

Reviewer: ```suggestion
    /// The time in microseconds of the texture since the start of the video. The timestamp is the video capture time of this texture in microseconds since the start of the video stream. The value does not have to be an exact capture time, but it should be larger than previous frame's timestamps. WebView2 will ignore a video frame if the timestamp is equal or less than the previous frame.
    /// This value will be exposed to script as the [`VideoFrame.timestamp` property](https://developer.mozilla.org/en-US/docs/Web/API/VideoFrame/timestamp).
```

    /// `value` is video capture time with microseconds units.
    /// The value does not have to be exact captured time, but it should be
    /// increasing order because renderer (composition) ignores incoming
    /// video frame (texture) if its timestamp is equal or prior to
    /// the current compositing video frame. It also will be exposed to the
    /// JS with `VideoFrame::timestamp`.
    /// (https://docs.w3cub.com/dom/videoframe/timestamp.html).

>     [propput] HRESULT Timestamp([out, retval] UINT64* value);

Reviewer: Let's include the unit in the property name.

```suggestion
    [propput] HRESULT TimestampInMicroseconds([out, retval] UINT64* value);
```

  //// corresponding event. There are no event args and the args
  //// parameter will be null.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);
}
/// This is the callback for stop request of texture stream.
[uuid(4111102a-d19f-4438-af46-efc563b2b9cf), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamStoppedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event. There are no event args and the args
  /// parameter will be null.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);
}
/// This is the callback for texture stream rendering error.
[uuid(52cb8898-c711-401a-8f97-3646831ba72d), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamErrorReceivedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] ICoreWebView2StagingTextureStreamErrorReceivedEventArgs* args);
}

> /// This is the event args interface for texture stream error callback.

Reviewer: ```suggestion
/// The event args for the `ICoreWebViewTextureStream ErrorReceived` event.
```

[uuid(0e1730c1-03df-4ad2-b847-be4d63adf700), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamErrorReceivedEventArgs : IUnknown {

>   /// Error kind.

Reviewer: ```suggestion
  /// The kind of error that has occurred.
```

  [propget] HRESULT Kind([out, retval]
      COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND* value);

>   /// texture that the error is associated with.

Reviewer: ```suggestion
  /// The texture with which this error is associated. For the `COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED` error kind, this property will be `nullptr`.
```

}
[uuid(431721e0-0f18-4d7b-bd4d-e5b1522bb110), object, pointer_default(unique)]
interface ICoreWebView2StagingRenderAdapterLUIDChangedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] IUnknown* args);
}
[uuid(431721e0-0f18-4d7b-bd4d-e5b1522bb110), object, pointer_default(unique)]
interface ICoreWebView2StagingRenderAdapterLUIDChangedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingEnvironment * sender,
      [in] IUnknown* args);
}
/// This is the callback for web texture.
[uuid(9ea4228c-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventHandler : IUnknown {
  /// Called to provide the implementer with the event args for the
  /// corresponding event.
  HRESULT Invoke(
      [in] ICoreWebView2StagingTextureStream* sender,
      [in] ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs* args);
}

> /// This is the event args interface for web texture.

Reviewer: ```suggestion
/// The event args for the `ICoreWebView2TextureStream WebTextureReceived` event.
```

[uuid(a4c2fa3a-295a-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2StagingTextureStreamWebTextureReceivedEventArgs : IUnknown {
  /// Return ICoreWebView2StagingWebTexture object.
  /// The call does not create new ICoreWebView2StagingWebTexture object, instead
  /// returns the same object.

>   /// The texture handle will be reused when ICoreWebView2StagingWebTexture

Reviewer: This seems like good info for the WebTexture's property documentation rather than this. Please remove it from here and I'll add it to the property docs in WebTexture.
